ingestion/tests/integration/mlflow/conftest.py

```diff
@@ -0,0 +1,176 @@
+
+
+# ------------------------------------------------------------
+# Container Configurations
+# ------------------------------------------------------------
+@dataclass
+class MySqlContainerConfigs:
+    """MySQL Configurations"""
+
+    image: str = "mysql:8"
+    username: str = "mlflow"
+    password: str = "password"
+    dbname: str = "experiments"
+    port: int = 3306
+    container_name: str = "mlflow-db"
+    exposed_port: Optional[int] = None
+
+    def with_exposed_port(self, container):
+        self.exposed_port = container.get_exposed_port(self.port)
+
+
+@dataclass
+class MinioContainerConfigs:
+    """MinIO Configurations"""
+
+    access_key: str = "minio"
+    secret_key: str = "password"
+    port: int = 9000
+    container_name: str = "mlflow-artifact"
+    exposed_port: Optional[int] = None
+
+    def with_exposed_port(self, container):
+        self.exposed_port = container.get_exposed_port(self.port)
+
+
+@dataclass
+class MlflowContainerConfigs:
+    """MLFlow Configurations"""
+
+    backend_uri: str = "mysql+pymysql://mlflow:password@mlflow-db:3306/experiments"
+    artifact_bucket: str = "mlops.local.com"
+    port: int = 6000
+    exposed_port: Optional[int] = None
+
+    def with_exposed_port(self, container):
+        self.exposed_port = container.get_exposed_port(self.port)
+
+
+class MlflowTestConfiguration:
+    """Responsible to hold all the configurations used by the test"""
+
+    def __init__(self):
+        self.mysql_configs = MySqlContainerConfigs()
+        self.minio_configs = MinioContainerConfigs()
+        self.mlflow_configs = MlflowContainerConfigs()
+
+
+# ------------------------------------------------------------
+# Fixture to setup the environment
+# ------------------------------------------------------------
+@pytest.fixture(scope="session")
+def mlflow_environment():
+    config = MlflowTestConfiguration()
+
+    docker_network = get_docker_network()
+
+    minio_container = get_minio_container(config.minio_configs)
+    mysql_container = get_mysql_container(config.mysql_configs)
+    mlflow_container = build_and_get_mlflow_container(config.mlflow_configs)
+
+    with docker_network:
+        minio_container.with_network(docker_network)
+        mysql_container.with_network(docker_network)
+        mlflow_container.with_network(docker_network)
+        with mysql_container, minio_container, mlflow_container:
+            # minio setup
+            minio_client = minio_container.get_client()
+            minio_client.make_bucket(config.mlflow_configs.artifact_bucket)
+
+            config.mysql_configs.with_exposed_port(mysql_container)
+            config.minio_configs.with_exposed_port(minio_container)
+            config.mlflow_configs.with_exposed_port(mlflow_container)
+
+            yield config
+
+
+# ------------------------------------------------------------
+# Utility functions
+# ------------------------------------------------------------
+def find_free_port():
```

Comment:
while this is tempting... it has potential to produce unstable results as we expend usage of services that use up ports..

better to try to bind and retry if it fails...

https://stackoverflow.com/a/2471039